src/compute-client/src/controller.rs

```diff
@@ -889,19 +898,225 @@ impl<T: ComputeControllerTimestamp> CollectionState<T> {
+    ) -> Self {
         let since = Antichain::from_elem(timely::progress::Timestamp::minimum());
-        let mut state = Self::new(since, Vec::new(), Vec::new());
+        let mut state = Self::new(
+            id,
+            since,
+            Vec::new(),
+            Vec::new(),
+            introspection_tx,
+            None,
+            None,
+        );
         state.log_collection = true;
         // Log collections are created and scheduled implicitly as part of replica initialization.
         state.scheduled = true;
         state
     }
 }
+
+/// Manages certain introspection relations associated with a collection. Upon creation, it adds
+/// rows to introspection relations. When dropped, it retracts its managed rows.
+///
+/// TODO: `ComputeDependencies` could be moved under this.
+#[derive(Debug)]
+struct CollectionIntrospection<T> {
+    /// The ID of the compute collection.
+    collection_id: GlobalId,
+    /// A channel through which introspection updates are delivered.
+    introspection_tx: crossbeam_channel::Sender<IntrospectionUpdates>,
+    /// Introspection state for `mz_materialized_view_refreshes`.
+    /// `Some` if it is a REFRESH MV.
+    refresh_introspection_state: Option<RefreshIntrospectionState<T>>,
+}
+
+impl<T> CollectionIntrospection<T> {
+    fn send(&self, introspection_type: IntrospectionType, updates: Vec<(Row, Diff)>) {
+        let result = self.introspection_tx.send((introspection_type, updates));
+
+        if result.is_err() {
+            // The global controller holds on to the `introspection_rx`. So when we get here that
+            // probably means that the controller was dropped and the process is shutting down, in
+            // which case we don't care about introspection updates anymore.
+            tracing::info!(
+                ?introspection_type,
+                "discarding introspection update because the receiver disconnected"
+            );
+        }
+    }
+}
+
+impl<T: ComputeControllerTimestamp> CollectionIntrospection<T> {
+    pub fn new(
+        collection_id: GlobalId,
+        introspection_tx: crossbeam_channel::Sender<IntrospectionUpdates>,
+        initial_as_of: Option<Antichain<T>>,
+        refresh_schedule: Option<RefreshSchedule>,
+        upper: Antichain<T>,
+    ) -> CollectionIntrospection<T> {
+        let self_ = CollectionIntrospection {
+            collection_id,
+            introspection_tx,
+            refresh_introspection_state: match (refresh_schedule, initial_as_of) {
+                (Some(refresh_schedule), Some(initial_as_of)) => Some(
+                    RefreshIntrospectionState::new(refresh_schedule, initial_as_of, upper),
+                ),
+                (Some(_refresh_schedule), None) => {
+                    soft_panic_or_log!("if we have a refresh_schedule, then it's an MV, so we should also have an initial_as_of");
+                    None
+                }
+                _ => None,
+            },
+        };
```

Comment:
Maybe more clearly like this?

```suggestion
        let refresh_introspection_state = match (refresh_schedule, initial_as_of) {
            (Some(refresh_schedule), Some(initial_as_of)) => Some(
                RefreshIntrospectionState::new(refresh_schedule, initial_as_of, upper),
            ),
            (Some(_refresh_schedule), None) => {
                soft_panic_or_log!("if we have a refresh_schedule, then it's an MV, so we should also have an initial_as_of");
                None
            }
            _ => None,
        };
        let self_ = CollectionIntrospection {
            collection_id,
            introspection_tx,
            refresh_introspection_state,
        };
```

Reply:
Hmm, yes, changed